### PR TITLE
Adds stop, start and stop_if_next_hour_is_imminent tasks

### PR DESCRIPTION
**Pull request by giorgiosironi**
Sample commands:
`./bldr start:elife-bot--end2end`
`./bldr stop:elife-bot--end2end`
`./bldr stop_if_next_hour_is_imminent:elife-bot--end2end`

The first two act on EC2 stop/state commands, bringing each node in the stack to the state `stopping` and then `stopped`; or to the state `pending` and then `running`.

The command `stop_if_next_hour_is_imminent` will be used by Alfred in a periodical task to stop nodes that have been running when they have outlived their usefulness. Nodes will be lazily started by builds when needed, but will not run for more than one hour (which is the minimum EC2 billing after a `stopped->running` transition.)

### CODE REVIEW

File: src/buildercore/utils.py [hunk 1]
```diff
@@ -224,11 +224,13 @@ def ymd(dt=None, fmt="%Y-%m-%d"):
         dt = datetime.now() # TODO: replace this with a utcnow()
     return dt.strftime(fmt)
 
-def die(assertion, msg):
+# TODO: this name is the opposite of assert, assertions read like "die if 1 == 1"
+# I'd try ensure(assertion, msg)
+def die(assertion, msg, *args):
```
Comment (giorgiosironi) [inline src/buildercore/utils.py]: thoughts on this name?

**Comment by lsh-0** [comment]
just to verify this for myself (we're ok):
https://aws.amazon.com/premiumsupport/knowledge-center/ec2-instance-hour-billing/

File: src/buildercore/utils.py (hunk 1 above)
Comment (lsh-0) [inline src/buildercore/utils.py]: `ensure` is good, use that.

File: src/lifecycle.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+@requires_aws_stack
+@echo_output
+def last_start_time(stackname):
+    return lifecycle.last_start_time(stackname)
+
+@task
+@requires_aws_stack
+@timeit
+def stop_if_next_hour_is_imminent(stackname, minimum_minutes='55'):
+    # TODO: can we write a description of the @task somewhere?
```
Comment (lsh-0) [inline src/lifecycle.py]: Like this?
http://docs.fabfile.org/en/1.12/api/core/tasks.html#fabric.tasks.Task

Comment (giorgiosironi) [inline src/lifecycle.py]: I didn't know the docstring was exposes in the task list, then I'll remove the comment as the description is there.

Comment (lsh-0) [inline src/lifecycle.py]: oh, I understand your comment now. yes - the docstring for the function with the `task` annotation appears alongside the task name (albeit, truncated) when one does `./bldr -l`

each of the decorators should be using the `functools.wraps` decorator that ensures the wrapper func gets the docstring and other attributes of the wrapped function.

Comment (giorgiosironi) [inline src/lifecycle.py]: All of them do, I can confirm.

File: src/buildercore/lifecycle.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    
+    e.g. {'i-6f727961': 'stopped'}"""
+    return {node.id:node.state for node in _nodes(stackname, node_ids=node_ids)}
+
+def _nodes(stackname, state=None, node_ids=None):
+    "returns list of ec2 instances data for a *specific* stackname"
+    conn = _connection(stackname)
+    return conn.get_only_instances(filters=_all_nodes_filter(stackname, state=state, node_ids=node_ids))
+
+def _all_nodes_filter(stackname, state, node_ids):
```
Comment (lsh-0) [inline src/buildercore/lifecycle.py]: could this logic be merged into `buildercore.core:find_ec2_instance` ?

Comment (giorgiosironi) [inline src/buildercore/lifecycle.py]: More or less. I'll try doing it by changing a few names to make them all consistent.

File: src/buildercore/core.py [hunk 4]
```diff
@@ -122,20 +122,25 @@ def connect_aws_with_stack(stackname, service):
...
     # http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html
-    filter_by_cluster = {
+    conn = connect_aws_with_stack(stackname, 'ec2')
+    return conn.get_only_instances(filters=_all_nodes_filter(stackname, state=state, node_ids=node_ids))
+
+def _all_nodes_filter(stackname, state, node_ids):
+    query = {
         'tag-key': ['Cluster', 'Name'],
         'tag-value': [stackname],
-        'instance-state-name': ['running'],
```
Comment (lsh-0) [inline src/buildercore/core.py]: the default state of 'running' doesn't appear to have been preserved.

Comment (giorgiosironi) [inline src/buildercore/core.py]: amended the last commit

**Comment by lsh-0** [comment]
cool - please merge once Alfred is happy

**Pull request merged**